### PR TITLE
MTG-750 Calculating bubblegum and account checksums

### PR DESCRIPTION
**Pull request by snorochevskiy**
This is a preliminary code review that doesn't intend to be me merged into the main branch, but aims to reveal problems before the development in completed.

In scope of review, I would like to present:
* the DB structure for bubblegum changes and epoch checksums
* the DB structure for account NFT changes 
* epoch and grand epoch checksums calculation
* account buckets and grand buckets calculation
* writing of bubblegum changes from the depth of bubblegum processor
* removing of bubblegum changes in scope of forks cleaning
* writing of account changes in account processor
* emitting of epoch/buckets checksums calculation on epoch change
* emitting of epoch/bucket -re-calculation on late data arrival

See the technical design:
https://github.com/adm-metaex/doc/wiki/Data-consistency-for-peer%E2%80%90to%E2%80%90peer-indexers

### CODE REVIEW

**Review by n00m4d — COMMENTED** [review]
LGTM! Only one small question at the moment

**Review by n00m4d — COMMENTED** [review]
Oh and also while was looking at fork cleaner got a thought that maybe we should sync with sequence consistency checker before calculating epoch hash, not to calculate epoch hash if we have gaps in tree sequences.

File: nft_ingester/src/consistency_calculator.rs [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+            let new_epoch_key = BubblegumEpochKey::new(current_tree.unwrap(), target_epoch);
+            storage
+                .bubblegum_epochs
+                .put(new_epoch_key, BUBBLEGUM_EPOCH_CALCULATING);
+
+            hasher = Hasher::default();
+        }
+        hasher.hash(&k);
+        hasher.hash(&v);
```
[the remark this answers is not in the document]
Comment (snorochevskiy) [reply nft_ingester/src/consistency_calculator.rs:128]: @n00m4d 
I assume that theoretically (not sure how high the probability is) we might have a situation, when on different peers there are bubblegum changes with same pubkey+slot+seq, but different signatures.
It you believe that this is impossible, then I can remove the signature from the bubblegum change entity (and from the hashing).

**Review by StanChe — COMMENTED** [review]
still have some reading to do, but sharing the feedback as some comments are pending for quite a while

**Pull request closed without merging**